src/main/java/org/opentestsystem/ap/ims/service/ContentSpecIdUtil.java

```diff
@@ -0,0 +1,153 @@
+     * @param claim claim 1 - 4
+     * @param target target information, which should be a number or letter depending on subject
+     * @param ccss common core state standard string
+     * @param domain domain code
+     *
+     * @return String with formatted ID, or empty string if ID cannot be produced.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public String formatId(String itemId,
+                           String subjectString,
+                           String gradeString,
+                           String claim,
+                           String target,
+                           String ccss,
+                           String domain) {
+
+        return formatId(itemId, subjectString, gradeString, claim, target, ccss, domain, null);
+    }
+
+    /**
+     * Formats the given components into an Enhanced format ID string. Logs warning if ID cannot be produced.
+     *
+     * @param itemId ID of item that defines these components, used in log messages
+     * @param subjectString subject either "Math" or "ELA".
+     * @param gradeString intended grade level as a string, K, HS, or any of 1 - 12
+     * @param claim claim 1 - 4
+     * @param target target information, which should be a number or letter depending on subject
+     * @param ccss common core state standard string
+     * @param domain domain code
+     * @param emphasis m or a/s, or blank.
+     *
+     * @return String with formatted ID, or empty string if ID cannot be produced.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public String formatId(String itemId,
+                           String subjectString,
+                           String gradeString,
+                           String claim,
+                           String target,
+                           String ccss,
+                           String domain,
+                           String emphasis) {
+
+        // If any core data is blank, we cannot create ID.
+        // Core data is subject, grade, claim, target, domain.
+        if (StringUtils.isAnyBlank(subjectString, gradeString, claim, target, domain)) {
+            log.info("Enhanced ID info: skipping creation of Enhanced ID for ID " + itemId +
+                    " because of missing data.");
+            return EMPTY;
+        }
+
+        try {
+            ContentSpecSubject subject = convertSubject(subjectString);
+            ContentSpecGrade grade = convertGrade(gradeString);
+
+            if (subject == ContentSpecSubject.UNSPECIFIED) {
+                throw new ValidationException(ErrorSeverity.INVALID, "Subject is invalid: " + subjectString);
+            }
+
+            if (grade == ContentSpecGrade.UNSPECIFIED) {
+                throw new ValidationException(ErrorSeverity.INVALID, "Grade is invalid: " + gradeString);
+            }
+
+            if (subject == MATH && isBlank(ccss)) {
+                throw new ValidationException(ErrorSeverity.INVALID, "CCSS cannot be blank for Math IDs");
```

Comment:
Why does TIMS need to know these errors?  This seems like a validation that should be within the enhanced id library.

Reply:
Same goes for all the validation errors in this class.

Reply:
The library has to stay consistent with the C-sharp code, so it can't error on invalid or unspecified grade. (It does error on invalid subjects, so I can remove that check. I didn't think of that.) The library also does not have a problem with creating enhanced IDs even if some parts are blank or missing. It is TIMS-specific logic to suppress the IDs in these cases. Conversely, in the test package converter, we create enhanced IDs even when only given subject, grade, and claim number. Technically, given a valid subject and grade, we can create a valid enhanced ID. Given an unspecified grade, we would create an ID like "M" or "E". That seems like a bug, but again, it's consistent with the C-sharp library.

Reply:
Ok. I misunderstood the block here.  Please ignore my comment.